packages/api-server/src/api-gateway.ts

```diff
@@ -0,0 +1,118 @@
+import * as msgpack from '@msgpack/msgpack';
+import * as assert from 'assert';
+import WebSocket from 'ws';
+import logger, { CustomLogger } from './logger';
+import { WebSocketMiddleware } from './websocket-connect';
+
+export { CustomLogger as Logger };
+
+export interface RpcRequest<T = unknown> {
+  version: string;
+  method: string;
+  params?: T;
+  id?: string | number | null;
+}
+
+export interface RpcError {
+  code: number;
+  message: string;
+}
+
+export interface RpcResponse<T = unknown> {
+  version: string;
+  result?: T;
+  more?: boolean;
+  error?: RpcError;
+  id: string | number;
+}
+
+export interface Sender<T = unknown> {
+  socket: WebSocket;
+  send(data: T): void;
+  end(data: T): void;
+  error(error: RpcError): void;
+}
+
+export type RpcHandler<Param = any, Result = unknown> = (
+  params: Param,
+  sender: Sender<Result>,
+) => Promise<Result | void> | Result | void;
+
+export default class ApiGateway {
+  middleware: WebSocketMiddleware = (socket, data, next) => {
+    this._onMessage(socket, data, next);
+  };
+
+  registerHandler(method: string, cb: RpcHandler): void {
+    this._rpcHandlers[method] = cb;
+    logger.info(`registered handler for "${method}"`);
+  }
+
+  getLogger(name: string): CustomLogger {
+    return logger.child({ tag: name });
+  }
+
+  private _rpcHandlers: Record<string, RpcHandler> = {};
+
+  private async _onMessage(
+    socket: WebSocket,
+    data: WebSocket.Data,
+    next: () => void,
+  ): Promise<void> {
+    assert.ok(data instanceof Buffer);
+    const req = msgpack.decode(data) as RpcRequest;
+    assert.strictEqual('0', req.version);
+
+    const buildResponse = (response: Partial<RpcResponse>) => {
+      return {
+        version: '0',
+        id: req.id,
+        ...response,
+      };
+    };
+
+    const sender: Sender = {
+      socket,
+      send: (data) =>
+        req.id !== undefined &&
+        req.id !== null &&
+        socket.send(msgpack.encode(buildResponse({ result: data, more: true }))),
+      end: (data) =>
+        req.id !== undefined &&
+        req.id !== null &&
+        socket.send(msgpack.encode(buildResponse({ result: data }))),
+      error: (error) =>
+        req.id !== undefined &&
+        req.id !== null &&
+        socket.send(msgpack.encode(buildResponse({ error }))),
```

Comment:
What do you think about creating a method for these restrictions and encodings?

Reply:
That will be nice but probably not crucial yet. These undefined and null should always be there for a rpc request and never there for a rpc notify, I think a stronger type checking is probably better in the long term.